### ISSUE TITLE
About Mask Autoencoder

### ISSUE THREAD

**Issue by ZK-Zhou** [SHI-Labs/Compact-Transformers#72]
Thank you for your work and code. Have you ever tried MAE like pre-training on CCT？ I have tried it initially, but it doesn't seem to work well. Could you give me some guidance？ Thanks！

**Comment by stevenwalton** [comment]
We have not tried this, nor some of the other pre-training techniques. We'd be curious about your results. Our paper is more about demonstrating that ViTs can work on small datasets without pre-training (of any form, including transfer learning) and can be effectively trained from scratch. Essentially we create a better "tokenization" and embedding for ViTs. The idea is to show that ViTs don't need big compute, big data, nor complex training schemes to be useful. 

That said, the same pre-training techniques should apply. Per MAE, you'd need to create a decoder network to do the pixel reconstruction. Typically AEs have symmetric encoders and decoders, but MAE doesn't -- encoder has a depth of 24 and decoder has depth of 8, they also have different embedding dims with the decoder being half the encoder's. If you'd like to perform the same pretraining as them I suggest carefully looking at [FAIR's repo](https://github.com/facebookresearch/mae). They have instructions for this pre-training. But most importantly, you'd need to build the decoder (which will later be discarded).

**Comment by ZK-Zhou** [comment]
Hi， thank you for your kind reply！ The main problems of pre-training using CCT are information leakage caused by the previous convolutional layer and information loss caused by the Maxpooling layer. So I am trying to improve CCT-MAE.
Have a nice day!

**Comment by cvg25** [comment]
Hey @stevenwalton and @ZK-Zhou, 
This thread, toghether with [this other issue](https://github.com/SHI-Labs/Compact-Transformers/issues/65), sparked in me the curiousity of whether SSL pretraining could yield even better results, and **yes, it does** 😄. In my latest work: [**Escaping the Big Data Paradigm in Self-Supervised Representation Learning**](https://openreview.net/pdf?id=6cHUucnYOk), we apply masked image modeling pretraining. 

Inspired by CCT, we propose to use a Convolutional Tokenizer instead of traditional Patch and Embed strategy. However, we make the convolutions sparse so that it is compatible with mask image modeling pretraining. We propose a pretraining SSL framework "similar" to MAE approach.

For your reference, on **Flowers102 torch split**, we achieve **97.7% top-1** through attentive probing on the pretrained frozen model. Fine-tuning could probably yield even better results. Of course, pretraining from scratch using only the Flowers102 dataset, not adding any external data.

I thought you would find it interesting,
Cheers,
Carlos

**Comment by stevenwalton** [comment]
Hey @cvg25! Thanks for sharing! 

This looks pretty interesting. Good job! I'm curious if you tried comparing a bit more directly with CCT and did some ablations or had some intermediate results in that direction. No worries if you didn't. Hope reviewing went well for you!

**Comment by cvg25** [comment]
Thanks for replying @stevenwalton 😃!!

On the comparison to CCT, indeed, the focus of my early experiments was to improve upon [@alihassanijr results in this thread](https://github.com/SHI-Labs/Compact-Transformers/issues/65). For context: VIT-Small (45.67 %), CCT-7 (63.15%), SCOTT-7(79.1%) (_ours_), MIM-JEPA pretrained frozen SCOTT-7 (96.9%) (_ours best_) top-1 on Flowers102.

The approach was to evolve CCT architecture to make it compatible with MIM task. The modifications:
- ConvTok: make convs sparse to prevent information leakage and mask vanishing while also maintaining the size of sequence length produced by patch-and-embed, to enable mask_token replacement before transformer blocks. Sparse convs just don’t compute when the central kernel value is 0, so for non-masked images the behaviour of our tokenizer and CCT is the same.
- SeqLen: I found that using an attentive head, i.e, a cross attention block to yield better results. Both on SL and when probing on frozen SSL pretrained features.
- Transformer block: for the ffn I used swiglu instead of mlp.

These changes made SCOTT a bit superior to CCT when SL training from scratch. However,  from an architecture perspective for SL, I consider them to be equivalent with the slight difference that SCOTT is compatible with MIM and CCT is not.

That’s why on Table 1 of my work I report SCOTT-7 SL (which I consider them equivalent to CCT) but enables a fairer comparison against MIM-JEPA SSL pretrained reports as the underlying architecture is the same.

All in all, every result on my paper that reports SL training should be seen as CCT performance. Indeed, I tried my best to make that clear in the text 😁.

Reviewing results however were really disappointing btw. From the reviewer’s comments I feel they vaguely red the paper nor understood the main focus: to escape the big data paradigm… being my first paper ever, it was a bit hard to process 😅. 

As future works, I’d like to see how SCOTT would perform on ImageNet which I dont have compute available to test. And more interesting: can the masking strategy be learned instead of heuristically generated? That would unlock better pretraining in the wild 👀.

btw, I’m open to discuss, colaborate or cowork 😁

**Comment by stevenwalton** [comment]
Thanks for the update. 

I just looked at the reviews. Sorry to see you couldn’t get much more feedback. It’s hard to publish, especially when you are limited in compute. CCT never got published and my StyleNAT has been facing this same compute issue. 

If I can be blunt, I’m not sure this paper is ready for ICLR yet, but I definitely think it’s workable. You have good work here and I think it contributes to the community, but a bit more needs to be done. (But my bar is usually lower than most) I think most reviewers won’t know the purpose of the Flowers dataset so you will need to motivate it more. I think it could also help you add CIFAR-10 for people to be more familiar with, but I understand that’s not exactly the problem being addressed here. If you can, adding ImageNet will help but I understand your compute constraints (could be beneficial to put in appendix if you can’t train to convergence). Same goes for a 100M model. We got those exact same complaints about why not use a giant pertained model and who would ever want to train from scratch lol. Don’t get discouraged, but there is room to improve here. It’s just unfortunate that reviewers may not offer much advice in how to actually improve. Reviewers these days are not going to know all the nuances so it gets much trickier to write. I’m of the same opinion as Yann LeCun, if it’s on ArXiV then it’s published.

The next 3 months are swamped for me but I’m pretty open to collaboration. You can find my email on my (github) blog (I’m going to close the issue but this doesn’t mean the conversation needs to stop, just about issue tracking)
